### PR TITLE
Update mempool after execValidateBlock and keep track of committed txs

### PR DESCRIPTION
**Pull request by larskuhtz**
This PR addresses two issues with the current Mempool:

1.  The mempool currently only deletes transactions that are included in a new block from the list of pending transactions in calls of `execNewBlock`. That means that transactions are removed from the pending set only on nodes that are mining.

    The PR adds the respective calls of

    ```Haskell
    mpaProcessFork memPoolAccess p
    mpaSetLastHeader memPoolAccess p
    ```

    to `execValidateBlock`. `p` is the parent header of the currently validate header. This isn't optimal. Ideally, the transactions in the current header should be removed, too. However, the implementation of these functions of the mempool accesses the payload db in order to obtain the transactions of the block. At the time these calls are made the current block is generally not yet available in the chain database. 

This could be fixed by explicitly deleting the transactions of the currently validated block from the set of pending transactions. However, the overhead for keeping the transactions a little longer is small and the logic is simpler this way.

2.  Transactions that are included in a block are deleted from the mempool only 1. when included in a new block or 2. when they expire. The latter can take up to 2 days. In the former case the following can happen:

    1. Node `A` validates Block `b` with tx `t` and deletes `t` from the set of pending tx
    2. Node `B` hasn't yet validated `b` and synchronizes its mempool with `A`. `t` gets added to the mempool of `A`.
    3. Node `B` validates Block `b` and deletes `t` from the set of pending tx
    4. Node `A` synchronizes its mempool with `B`. `t` gets added to the mempool of `B`.
    5. At this point `A` and `B` both have validated `b` and have `t` in their mempools. It can take up to 2 days until `t` is removed.

    The PR introduces a probabilistic data structure that keeps track of unexpired transactions that have been included
    in blocks and prevents these from being added to the mempool.

TODO:

* [x] make `maxCurrentTxIdSize` configurable.
* [x] motivate PR and document changes
* [x] add code comment explaining why mempool update is done only for the parent and the current header
* [x] consider renaming `CurrentTxIdx` into something more meaningful
* [x] Provide telemetry and test results from mainnet tests.

### CODE REVIEW

File: src/Chainweb/Mempool/CurrentTxs.hs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+-- import Chainweb.Mempool.InMemTypes
+import Chainweb.Mempool.Mempool
+import Chainweb.Time
+import Chainweb.Utils
+
+-- -------------------------------------------------------------------------- --
+-- Configuration
+
+-- | Limits the size of the commit log. If the size is reached elements are
```
Comment (larskuhtz) [inline src/Chainweb/Mempool/CurrentTxs.hs:56]:
```suggestion
-- | Limits the size of the current txs set. If the size is reached elements are
```

File: src/Chainweb/Mempool/CurrentTxs.hs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+-- deleted, and queries may return false negatives. Not guarantee is made about
```
Comment (larskuhtz) [inline src/Chainweb/Mempool/CurrentTxs.hs:57]:
```suggestion
-- deleted, and queries may return false negatives. No guarantee is made about
```

File: src/Chainweb/Mempool/CurrentTxs.hs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+-- what elements are deleted.
```
Comment (larskuhtz) [inline src/Chainweb/Mempool/CurrentTxs.hs:58]:
```suggestion
-- | Limits the size of the set of current txs. If the size is reached elements are
-- deleted, and queries may return false negatives. No guarantee is made about
-- which elements are deleted.
```

**Comment by larskuhtz** [comment]
This PR has been tested for over a week on us-w2 and a common ubuntu node. Both nodes perform fine. Telemetry indicates that the PR works as expected.

File: src/Chainweb/Mempool/CurrentTxs.hs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        -- @maxCurrentTxsSize@ should be much larger than 1000. The worst case
+        -- expected running time is for @1000 == maxCurrentTxsSize@, which still
+        -- is very fast.
+        --
+        let n = min l (max 1000 (l - int maxCurrentTxsSize))
+        in take n . L.nub . randomRs (0, l-1) <$> newStdGen
+      else return []
+
+    -- Not sure if foldl' would be faster here
+    return $ CurrentTxs $! foldr S.deleteAt s0 indexesToBeDeleted
```
Comment (emilypi) [inline src/Chainweb/Mempool/CurrentTxs.hs:154]: It's probably good to `foldl'` either way - less unnecessary memory pressure.

**Pull request merged**